packages/core/src/shared/utilities/resourceCache.ts

```diff
@@ -0,0 +1,127 @@
+/*!
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import globals from '../extensionGlobals'
+import { globalKey } from '../globalState'
+import { getLogger } from '../logger/logger'
+import { waitUntil } from '../utilities/timeoutUtils'
+
+interface Resource<V> {
+    result: V | undefined
+    locked: boolean
+    timestamp: number
+}
+
+// GlobalStates schema, which is used for vscode global states deserialization
+// [globals.globalState.tryGet<T>]
+export interface GlobalStateSchema<V> {
+    resource: Resource<V>
```

Comment:
why is in this indirection needed?

Reply:
not sure if i get your question correctly justin.

i was thinking and based on the comments in the previous PR #7151, we might want to put related things/fields under the same global key as much as we can but not only the resource itself. So here i need a super type which contains the resource but also callers can add whatever they want here and this class shouldn't touch field other than "resource".

For my own use case i want to store (1) resource (2) map of connectionId to previousSelectedProfile
```
interface RegionProfileGlobalState extends GlobalStateSchema<RegionProfile> {
    previousSelection: { [label: string]: RegionProfile }
}
```

does this make sense to you?

Reply:
btw appreciate your early feedback, im not a typescript expert, your feedback would be very helpful

Reply:
> For my own use case i want to store (1) resource (2) map of connectionId to previousSelectedProfile

```
interface RegionProfileGlobalState extends GlobalStateSchema<RegionProfile> {
    previousSelection: { [label: string]: RegionProfile }
}
```

as discussed in the previous PR, i won't use the same global key this time for not breaking existing users' experience

Reply:
so i will revert https://github.com/aws/aws-toolkit-vscode/pull/7151/commits/3e501456fb1cd7b125819db2fbe7390cdfdc2558 partially